### PR TITLE
Allow user to constrain supported lock environments

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

The strategy here is: if the user provides supported environments, we use those as the initial forks when resolving. As a result, we never add or explore branches that are disjoint with the supported environments. (If the supported environments change, we ignore the lockfile entirely, so we don't have to worry about any interactions between supported environments and the preference forks.)

Closes https://github.com/astral-sh/uv/issues/6184.

### CODE REVIEW

File: crates/uv-resolver/src/lock.rs [hunk 1]
```diff
@@ -951,6 +979,8 @@ struct LockWire {
     /// forks in the lockfile so we can recreate them in subsequent resolutions.
     #[serde(rename = "environment-markers", default)]
     fork_markers: Vec<MarkerTree>,
+    #[serde(rename = "supported-markers", default)]
+    supported_environments: Vec<MarkerTree>,
```
Comment (charliermarsh) [inline crates/uv-resolver/src/lock.rs:983]: I am tempted to rename these to: `resolution-markers` and `supported-markers`.

File: crates/uv/tests/lock.rs [hunk 2]
```diff
@@ -9918,3 +9918,299 @@ fn lock_exclude_unnecessary_python_forks() -> Result<()> {
...
+        [[package]]
+        name = "black"
+        version = "24.3.0"
+        source = { registry = "https://pypi.org/simple" }
+        dependencies = [
+            { name = "click", marker = "platform_system != 'Windows'" },
+            { name = "mypy-extensions", marker = "platform_system != 'Windows'" },
+            { name = "packaging", marker = "platform_system != 'Windows'" },
+            { name = "pathspec", marker = "platform_system != 'Windows'" },
+            { name = "platformdirs", marker = "platform_system != 'Windows'" },
```
Comment (charliermarsh) [inline crates/uv/tests/lock.rs:9980]: I think there's another solution where we don't treat the `supported-markers` as the input forks, and instead just test for disjointness everywhere (which would avoid having to write these markers on dependencies). But this is much simpler to implement, and we can always change the strategy later.

**Review by konstin — APPROVED** [review]

File: crates/uv/src/commands/project/lock.rs [hunk 3]
```diff
@@ -462,19 +465,20 @@ async fn do_lock(
...
-                    .map(|existing_lock| existing_lock.lock().fork_markers().to_vec())
-                    .unwrap_or_default()
-            });
+            let resolver_markers = ResolverMarkers::universal(
+                reusable_lock
+                    .map(|lock| lock.fork_markers().to_vec())
+                    .unwrap_or_else(|| {
+                        environments
+                            .cloned()
+                            .map(SupportedEnvironments::into_markers)
```
Comment (konstin) [inline crates/uv/src/commands/project/lock.rs:478]: This is actually critical: iirc the resolver currently assumes these are non-overlapping, and if they aren't we can create a conflicting resolution

Comment (charliermarsh) [reply crates/uv/src/commands/project/lock.rs:478]: Wait, what? Is that true? Where does that assumption come into play?

Comment (charliermarsh) [reply crates/uv/src/commands/project/lock.rs:478]: I think you're right that they need to be disjoint. We can either (1) require that the user provides disjoint markers, or (2) `and` with the negation of the preceding markers when constructing this.

**Comment by zanieb** [comment]
Ideally would be documented in the project or resolver concept guide too.

File: crates/uv-workspace/src/environments.rs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    /// Convert the [`SupportedEnvironments`] struct into a list of marker trees.
+    pub fn into_markers(self) -> Vec<MarkerTree> {
+        self.0
+    }
+}
+
+/// Serialize a [`SupportedEnvironments`] struct into a list of marker strings.
+impl serde::Serialize for SupportedEnvironments {
```
[the remark this answers is not in the document]
Comment (charliermarsh) [reply crates/uv-workspace/src/environments.rs:23]: We support providing a string or a list of strings -- either is fine.

File: crates/uv/tests/lock.rs [hunk 5]
```diff
@@ -9918,3 +9918,299 @@ fn lock_exclude_unnecessary_python_forks() -> Result<()> {
...
+    ----- stderr -----
+    warning: `uv lock` is experimental and may change without warning
+    Resolved 7 packages in [TIME]
+    "###);
+
+    let lock = fs_err::read_to_string(context.temp_dir.join("uv.lock")).unwrap();
+
+    // Because we're _not_ locking for Windows, `colorama` should not be included.
+    insta::with_settings!({
+        filters => context.filters(),
```
[the remark this answers is not in the document]
Comment (charliermarsh) [reply crates/uv/tests/lock.rs:9955]: No, shouldn't be.

**Comment by charliermarsh** [comment]
Ok, make it require disjoint forks.

**Comment by charliermarsh** [comment]
Also added docs.

File: crates/uv/src/commands/project/lock.rs (hunk 3 above)
Comment (konstin) [reply crates/uv/src/commands/project/lock.rs:513]: Putting this here also for reference: The problem is: If we resolve foo==1 for `(sys_platform == "linux" or sys_platform == "darwin")` and foo==2 for `(sys_platform == "linux" or sys_platform == "nt")`, which one would we install on linux?

File: docs/concepts/projects.md [hunk 6]
```diff
@@ -63,8 +63,21 @@ use [`uvx`](../guides/tools.md) or
...
 
-`uv.lock` is a _universal_ lockfile that captures the packages that would be installed across all
-possible Python markers such as operating system, architecture, and Python version.
+`uv.lock` is a _universal_ or _cross-platform_ lockfile that captures the packages that would be
+installed across all possible Python markers such as operating system, architecture, and Python
+version.
+
+If your project supports a more limited set of platforms or Python versions, you can constrain the
+set of solved platforms via the `environments` setting, which accepts a list of PEP 508 environment
+markers. For example, to constrain the lockfile to macOS and Linux, and exclude Windows:
```
Comment (zanieb) [inline docs/concepts/projects.md:72]: Any note that these need to be disjoint?

Comment (zanieb) [reply docs/concepts/projects.md:72]: (and what that means)

**Review by BurntSushi — APPROVED** [review]

**Pull request merged**